File: src/main/java/org/kite9/diagram/position/CostedDimension.java
package org.kite9.diagram.position;

import org.kite9.diagram.adl.DiagramElement;

/**
 * This extends the idea of dimension, but allows you to associate a cost with the dimension
 * other than simply the size.
 * 
 * @author robmoffat
 *
 */
public class CostedDimension extends Dimension2D {

	private static final long serialVersionUID = 5835848361276316308L;

	public static final CostedDimension ZERO = new CostedDimension(0,0,0) {

		private static final long serialVersionUID = -5465011727879693143L;
		
	};
	
	public static final CostedDimension NOT_DISPLAYABLE = new CostedDimension(-1,-1,Integer.MAX_VALUE);
	
	public static final CostedDimension UNBOUNDED = new CostedDimension(Integer.MAX_VALUE,Integer.MAX_VALUE,0);
	

	public long cost;

	
	public CostedDimension() {
		super();
	}
	
	public CostedDimension(Dimension2D d) {
		super(d.getWidth(), d.getHeight());
	}

	public CostedDimension(double arg0, double arg1, long cost) {
		super(arg0, arg1);
		this.cost = cost;
	}

	/**
	 * Works out cost based on how well the new {@link CostedDimension} fits into within.
	 */
	public CostedDimension(double width, double height, Dimension2D within) {
		super(width, height);
		if (within != null) {
		double extraHeight = Math.max(height - within.getHeight(), 0);
		double extraWidth = Math.max(width - within.getWidth(), 0);
		
		cost = (long)( (extraHeight * width) + (extraWidth * height) +  (extraHeight * extraWidth));
		}
	}

	public long getCost() {
		return cost;
	}

	public void setCost(int cost) {
		this.cost = cost;
	}
	
	public static CostedDimension chooseBest(CostedDimension a, CostedDimension b) {
		if (a == CostedDimension.NOT_DISPLAYABLE) {
			if (b == CostedDimension.NOT_DISPLAYABLE) {
				return CostedDimension.NOT_DISPLAYABLE;
			}
			return b;
		} else {
			if (a.cost < b.cost) {
				return a;
			} else if (a.cost > b.cost) {
				return b;
			}
			
			if (a.getHeight()< b.getHeight()) {
				return a;
			} else if (a.getHeight() > b.getHeight()) {
				return b;
			}
		
			if (a.getWidth()< b.getWidth()) {
				return a;
			} else if (a.getWidth() > b.getWidth()) {
				return b;
			}
	
			return a;
		}
	}

	public int compareTo(DiagramElement o) {
		if (o instanceof CostedDimension) {
			return ((Long)this.cost).compareTo(((CostedDimension)o).cost);
		} else {
			return 0;
		}
	}
}

File: src/main/kotlin/org/kite9/diagram/model/position/CostedDimension.kt
package org.kite9.diagram.model.position

import org.kite9.diagram.model.DiagramElement

/**
 * This extends the idea of dimension, but allows you to associate a cost with the dimension
 * other than simply the size.
 *
 * @author robmoffat
 */
open class CostedDimension : Dimension2D {
    var cost: Long = 0

    constructor() : super() {}
    constructor(d: Dimension2D) : super(d.width, d.height) {}
    constructor(arg0: Double, arg1: Double, cost: Long) : super(arg0, arg1) {
        this.cost = cost
    }

    /**
     * Works out cost based on how well the new [CostedDimension] fits into within.
     */
    constructor(width: Double, height: Double, within: Dimension2D?) : super(width, height) {
        if (within != null) {
            val extraHeight = Math.max(height - within.height, 0.0)
            val extraWidth = Math.max(width - within.width, 0.0)
            cost = (extraHeight * width + extraWidth * height + extraHeight * extraWidth).toLong()
        }
    }

    fun setCost(cost: Int) {
        this.cost = cost.toLong()
    }

    operator fun compareTo(o: DiagramElement?): Int {
        return if (o is CostedDimension) {
            cost.compareTo((o as CostedDimension).cost)
        } else {
            0
        }
    }

    companion object {

		val ZERO: CostedDimension = CostedDimension(0.0, 0.0, 0)
        val NOT_DISPLAYABLE: CostedDimension = CostedDimension(-1.0, -1.0, Long.MAX_VALUE)
		val UNBOUNDED: CostedDimension = CostedDimension(Double.MAX_VALUE, Double.MAX_VALUE, 0)

        fun chooseBest(a: CostedDimension, b: CostedDimension): CostedDimension {
            return if (a === NOT_DISPLAYABLE) {
                if (b === NOT_DISPLAYABLE) {
                    NOT_DISPLAYABLE
                } else b
            } else {
                if (a.cost < b.cost) {
                    return a
                } else if (a.cost > b.cost) {
                    return b
                }
                if (a.height < b.height) {
                    return a
                } else if (a.height > b.height) {
                    return b
                }
                if (a.width < b.width) {
                    return a
                } else if (a.width > b.width) {
                    return b
                }
                a
            }
        }
    }
}